File: 0064_minimum_path_sum/min_path_sum.cpp
#include <iostream>
#include <climits>
#include <vector>

using namespace std;

/**
 * Time: O(m * n)
 * Space: O(1)
 */
int min_path_sum(vector<vector<int>> &grid)
{
    int r = grid.size();
    if (r == 0)
        return 0;
    int c = grid[0].size();
    for (int i = 0; i < r; i++)
    {
        for (int j = 0; j < c; j++)
        {
            if (!i && !j)
                continue;
            int left = !j ? INT_MAX : grid[i][j - 1];
            int up = !i ? INT_MAX : grid[i - 1][j];
            grid[i][j] += min(left, up);
        }
    }
    return grid[r - 1][c - 1];
}

File: 0064_minimum_path_sum/min_path_sum.py
from typing import List


def min_path_sum(grid: List[List[int]]) -> int:
    """
    Time: O(m * n)
    Space: O(m * n)
    """
    r, c = len(grid), len(grid[0])
    dp = [[0 for _ in range(c)] for _ in range(r)]
    dp[0][0] = grid[0][0]
    for i in range(1, r):
        dp[i][0] = dp[i-1][0] + grid[i][0]
    for i in range(1, c):
        dp[0][i] = dp[0][i-1] + grid[0][i]
    for i in range(1, r):
        for j in range(1, c):
            dp[i][j] = grid[i][j] + min(dp[i-1][j], dp[i][j - 1])
    return dp[-1][-1]
